storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -2156,11 +2156,111 @@ static bool innodb_init_param(void) {
   return (TRUE);
 }
 
+/** Scan the sdi id from mysql.tables
+@param[in]	name 		Name of table database/name
+@param[out]     sdi_id		id of table
+@param[in]	thd 		THD */
+void get_id_from_mysql_table_scan(const std::string name, uint64 *id,
+                                  THD *thd) {
+  std::string db_name;
+  std::string table_name;
+  uint64 schema_id = 0;
+
+  /* get the database and table_name from name */
+  dict_name::get_table(name, db_name, table_name);
+
+  ut_ad(db_name.compare("mysql") != 0);
+
+  dict_table_t *sys_tables = nullptr;
+  btr_pcur_t pcur;
+  const rec_t *rec = nullptr;
+  mtr_t mtr;
+  MDL_ticket *mdl = nullptr;
+  mem_heap_t *heap = mem_heap_create(1000);
+
+  mutex_enter(&(dict_sys->mutex));
+
+  auto build_mysql_schema = [&]() {
+    std::map<std::string, int> mysql_schema;
+
+    mtr_start(&mtr);
+    rec = dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/schemata",
+                              &sys_tables);
+    while (rec) {
+      uint64 rec_schema_id;
+      std::string rec_name;
+
+      dd_process_schema_rec(heap, rec, sys_tables, &mtr, &rec_name,
+                            &rec_schema_id);
+      mysql_schema.insert(std::pair<std::string, int>(rec_name, rec_schema_id));
+      mem_heap_empty(heap);
+
+      mtr_start(&mtr);
+      rec = (rec_t *)dd_getnext_system_rec(&pcur, &mtr);
+    }
+
+    mtr_commit(&mtr);
+    dd_table_close(sys_tables, thd, &mdl, true);
+    mem_heap_empty(heap);
+
+    return mysql_schema;
+  };
+
+  static auto mysql_schema = build_mysql_schema();
+
+  schema_id = mysql_schema[db_name];
+
+  ut_ad(schema_id != 0);
+  if (schema_id == 0) {
+    msg("xtrabackup: can't find %s entry in mysql/schemata for space %s\n",
```

Comment:
for tablespace

Reply:
I know that you added exit() but an mtr_commit() would be good here. its like not closing a file and exiting. it may be ok but better to follow proper cleanups/exits

also, am not a fan of exit() calls. Better to do proper exit of threads, propagation of errors, cleanups, and then return. This you can ignore though [ follow original style ]

Reply:
mtr_commit is not required here. for first call it is done just above  dd_table_close. let me follow orignal style.